lib/tapioca/dsl/compilers/json_api_client_resource.rb

```diff
@@ -0,0 +1,208 @@
+      #     def user_id=(user_id); end
+      #   end
+      # end
+      # ~~~
+      class JsonApiClientResource < Compiler
+        extend T::Sig
+
+        ConstantType = type_member { { fixed: T.class_of(::JsonApiClient::Resource) } }
+
+        sig { override.void }
+        def decorate
+          schema = resource_schema
+          return if schema.nil? && constant.associations.empty?
+
+          root.create_path(constant) do |k|
+            module_name = "JsonApiClientResourceGeneratedMethods"
+            k.create_module(module_name) do |mod|
+              schema&.each_property do |property|
+                generate_methods_for_property(mod, property)
+              end
+
+              constant.associations.each do |association|
+                generate_methods_for_association(mod, association)
+              end
+            end
+
+            k.create_include(module_name)
+          end
+        end
+
+        class << self
+          extend T::Sig
+
+          sig { override.returns(T::Enumerable[Module]) }
+          def gather_constants
+            all_modules.select do |c|
+              name_of(c) && c < ::JsonApiClient::Resource
+            end
+          end
+        end
+
+        private
+
+        sig { returns(T.nilable(::JsonApiClient::Schema)) }
+        def resource_schema
+          schema = constant.schema
+
+          # empty? does not exist on JsonApiClient::Schema
+          schema if schema.size > 0 # rubocop:disable Style/ZeroLengthPredicate
+        end
+
+        sig do
+          params(
+            mod: RBI::Scope,
+            property: ::JsonApiClient::Schema::Property,
+          ).void
+        end
+        def generate_methods_for_property(mod, property)
+          type = type_for(property)
+
+          name = property.name.to_s
+
+          mod.create_method(name, return_type: type)
+          mod.create_method("#{name}=", parameters: [create_param(name, type: type)], return_type: type)
+        end
+
+        sig { params(property: ::JsonApiClient::Schema::Property).returns(String) }
+        def type_for(property)
+          type = ::JsonApiClient::Schema::TypeFactory.type_for(property.type)
+          return "T.untyped" if type.nil?
+
+          sorbet_type = if type.respond_to?(:sorbet_type)
+            type.sorbet_type
+          elsif type == ::JsonApiClient::Schema::Types::Integer
+            "::Integer"
+          elsif type == ::JsonApiClient::Schema::Types::String
+            "::String"
+          elsif type == ::JsonApiClient::Schema::Types::Float
+            "::Float"
+          elsif type == ::JsonApiClient::Schema::Types::Time
+            "::Time"
+          elsif type == ::JsonApiClient::Schema::Types::Decimal
+            "::BigDecimal"
+          elsif type == ::JsonApiClient::Schema::Types::Boolean
+            "T::Boolean"
+          else
+            "T.untyped"
+          end
+
+          if property.default.nil?
+            as_nilable_type(sorbet_type)
```

Comment:
Properties with default values are always initialized with their default value, even when creating a bare object.

We made the decision to make those properties non-nilable to make the usage cleaner.

That being said, it _is_ possible today to have a nil:

```rb
class User < JsonApiClient::Resource
  property :name, type: :string, default: "foo"
end

> u = User.new
=> #<User:@attributes={"type"=>"users", "name"=>"foo"}>

> u.name
=> "foo"

> u.name = 123
=> 123

> u.name # Casts to string
=> "123"

> u.name = nil

> u.name # Actually nil
=> nil
```